Mathlib/MeasureTheory/Measure/SeparableMeasureSpace.lean

```diff
@@ -0,0 +1,455 @@
+        (μ := μ.restrict (T N)) h𝒜 hgen).approx s ms
+        (ne_of_lt (lt_of_le_of_lt (μ.restrict_apply_le _ s) hμs.lt_top))
+        (ε / 2) (by linarith [ε_pos])
+        with ⟨t, t_mem, ht⟩
+      -- We can then use `t ∩ (S N)`, because `S N ∈ 𝒜` by hypothesis.
+      -- `μ (s ∆ (t ∩ S N))`
+      --   `≤ μ (s ∆ (s ∩ S N)) + μ ((s ∩ S N) ∆ (t ∩ S N))`
+      --   `= μ s - μ (s ∩ S N) + μ (s ∆ t) ∩ S N) < ε`.
+      refine ⟨t ∩ T N, h𝒜.inter_mem t_mem (T_mem N), ?_⟩
+      calc
+        μ (s ∆ (t ∩ T N))
+          ≤ μ (s \ (s ∩ T N)) + μ ((s ∆ t) ∩ T N) := by
+              rw [← symmDiff_of_le (inter_subset_left ..), symmDiff_comm _ s,
+                inter_symmDiff_distrib_right]
+              exact measure_symmDiff_le _ _ _
+        _ < ENNReal.ofReal (ε / 2) + ENNReal.ofReal (ε / 2) := by
+              apply ENNReal.add_lt_add
+              · rw [measure_diff
+                    (inter_subset_left ..)
+                    (ms.inter (hgen ▸ measurableSet_generateFrom (T_mem N)))
+                    (ne_top_of_le_ne_top hμs (measure_mono (inter_subset_left ..))),
+                  lt_ofReal_iff_toReal_lt (sub_ne_top hμs),
+                  toReal_sub_of_le (measure_mono (inter_subset_left ..)) hμs]
+                apply lt_of_le_of_lt (sub_le_dist ..)
+                nth_rw 1 [← univ_inter s]
+                rw [inter_comm s, dist_comm, ← T_spanning, iUnion_inter _ T]
+                apply hN N (le_refl _)
+              · rwa [← μ.restrict_apply' (hgen ▸ measurableSet_generateFrom (T_mem N))]
+        _ = ENNReal.ofReal ε := by
+              rw [← ofReal_add (by linarith [ε_pos]) (by linarith [ε_pos]), add_halves]
+    · exact fun n ↦ ne_top_of_le_ne_top hμs (measure_mono (inter_subset_right ..))
+    · exact ne_top_of_le_ne_top hμs
+        (measure_mono (iUnion_subset (fun i ↦ inter_subset_right ..)))
+
+end MeasureDense
+
+section SeparableMeasureSpace
+
+/-! ### Definition of a separable measure space, sufficient condition -/
+
+/-- A measure space `X` is separable if it admits a countable and measure-dense family of sets.
+
+The term "separable" is justified by the fact that the definition translates to the usual notion
+of separability in the metric space made by constant indicators equipped with the `L¹` norm. -/
+class SeparableMeasureSpace (μ : Measure X) : Prop :=
+  exists_countable_measureDense : ∃ 𝒜, 𝒜.Countable ∧ μ.MeasureDense 𝒜
+
+/-- By definition, a separable measure space admits a countable and measure-dense family of sets. -/
+theorem exists_countable_measureDense [SeparableMeasureSpace μ] :
+    ∃ 𝒜, 𝒜.Countable ∧ μ.MeasureDense 𝒜 :=
+  SeparableMeasureSpace.exists_countable_measureDense
+
+/-- If a measurable space is countably generated and equipped with a `σ`-finite measure, then it
+is separable. -/
+instance instSeparableMeasureSapaceCountablyGeneratedSigmaFinite [CountablyGenerated X]
+    [SigmaFinite μ] : SeparableMeasureSpace μ where
```

Comment:
Is it possible to weaken `SigmaFinite` to `SFinite` here?

Reply:
I don't think so. Consider the infinite sum of the Lebesgue measure on $[0,1]$. Then $\mathcal{A}$ is measure-dense if for any measurable set $s$ there exists $t \in \mathcal{A}$ such that $\mu(s \Delta t) = 0$. Now consider countable unions of intervals of the form $[1/(n+1), 1/n]$. This gives an uncountable family $\mathcal{B}$ of measurable sets such that the symmetric difference of two of those sets is positive (so infinite in this case). Moreover if $\mu(s \Delta t) = 0$ and $\mu(s \Delta s') = \infty$ then $\mu(s' \Delta t) = \infty$, so $t \in \mathcal{A}$ can only approximate at most one set in $\mathcal{B}$, so $\mathcal{A}$ cannot be countable.

Reply:
Doesn't measure-dense only require approximation for finite-measure sets? In your example, all the sets have infinite measure, so they wouldn't need to be approximated by elements of \mathcal{A}, right?

Reply:
Woops sorry I had forgotten my own definition!

Reply:
Ok after digging the internet I think I managed to prove it but it is not straight-forward: $\mu$ can be written $\mu = \mu_{sf} + \mu_{0-\infty}$, where $\mu_{sf}$ is semi-finite (any set with infinite measure has a subset with positive finite measure) and $\mu_{0-\infty}$ takes values $0$ or $\infty$. $\mu_{0-\infty}$ is separable. If $\mu$ is s-finite, then $\mu_{sf}$ is still s-finite I think (we remove the support of $\mu_{0-\infty}$), and then I think a measure which is both s-finite and semi-finite is $\sigma$-finite, because it is equivalent to a $\sigma$-finite measure and its density has to be finite everywhere, therefore it is separable. The sum of two separable measures is separable. I am not totally sure, I have never manipulated s-finite and semi-finite measures before.
I did not manage to do the proof directly however, and it looks like semi-finite measures are not in Mathlib yet.

I even think we can prove that separable measures on a countably-generated space are exactly measures which are the sum of a $\sigma$-finite measure and a $(0-\infty)$-measure.

Reply:
I have on a branch that every s-finite measure can be written as the sum of a sigma-finite measure and a measure which takes only values 0 and infinity. I will be PRing it in the coming days.
However I am also busy with other things and if you wait for my result it will probably delay this PR by at least a week.

Reply:
Oh nice! I don't know what @sgouezel thinks, the goal of this PR was mostly second-countability of basic Lp spaces (so with Lebesgue measure), so I think I can add the generalization to s-finite measures in a future PR.

Reply:
It can definitely wait for an improvement to S-finite measures in a future PR. Maybe add a TODO about this?

Reply:
In fact #12759 has already been merged, and I think it contains everything you need to extend your result to s-finite measures (see in particular the instance saying that `SigmaFinite (μ.restrict μ.sigmaFiniteSet)`, and `measure_eq_zero_or_top_of_subset_compl_sigmaFiniteSet`).

Reply:
Thanks, I'll look into it.

Reply:
I added the generalization to s-finite measures!